### PR TITLE
DDP for FineTuning

### PR DESCRIPTION
**Pull request by psinger-prior**
## Issue
Closing https://github.com/PriorLabs/TabPFN/issues/809

## Motivation and Context

---

## Public API Changes

-   [x] No Public API changes
-   [ ] Yes, Public API changes (Details below)

---

## How Has This Been Tested?

Running example scripts on single and multi gpu nodes.

---

## Checklist

-   [x] The changes have been tested locally.
-   [x] Documentation has been updated (if the public API or usage changes).
-   [x] A changelog entry has been added (see `changelog/README.md`), or "no changelog needed" label requested.
-   [x] The code follows the project's style guidelines.
-   [x] I have considered the impact of these changes on the public API.

---

### CODE REVIEW

**Comment by alanprior** [comment]
@psinger-prior sorry, I've completely missed this PR, probably due to the war. @anuragg1209 would you feel OK to review it?

**Comment by anuragg1209** [comment]
> @psinger-prior sorry, I've completely missed this PR, probably due to the war. @anuragg1209 would you feel OK to review it?

Hi @alanprior, yes, this PR review is under my To-Dos. Please feel free to unsubscribe.

**Comment by anuragg1209** [comment]
Hi @psinger-prior,
Thanks for the PR. I tested the DDP on 4 H100s and everything worked correctly. I wrote test scripts for classifier DDP, regressor DDP, and gradient synchronization verification (parameter difference across ranks: 0.00e+00). All the tests passed, and the core implementation looks solid.

I just have a few comments to add.

File: src/tabpfn/finetuning/finetuned_base.py [hunk 1]
```diff
@@ -720,47 +935,64 @@ def _fit(  # noqa: C901,PLR0912
...
-            self.finetuned_estimator_ = best_model
+        if self.early_stopping and best_model_state is not None:
+            self.finetuned_estimator_.model_.load_state_dict(best_model_state)
+
+        # --- DDP cleanup ---
+        self._ddp_module_ = None
+        if using_ddp:
+            dist.barrier()
+            if created_process_group:
+                dist.destroy_process_group()
```
Comment (anuragg1209) [inline src/tabpfn/finetuning/finetuned_base.py:992]: **Edge case: Multiple fit() calls in one torchrun session can deadlock**

- After `fit()` completes, `dist.destroy_process_group()` is called in `finetuned_base.py`
- If the user calls `fit()` again on a new model, all ranks must re-enter `dist.init_process_group()` at line 79 simultaneously                                                              
  - But rank 0 may be delayed in doing prediction/evaluation (only rank 0 runs `predict()`), while rank 1 already reaches `init_process_group()` and blocks waiting                               
- Result: deadlock until the 30-min NCCL timeout. 

Potential Remedy:
Add a warning in the docs/examples: "Do not call fit() on multiple finetuning instances within a single torchrun session. Use separate scripts or separate torchrun invocations."  

Wdyt?

File: src/tabpfn/finetuning/finetuned_base.py [hunk 2]
```diff
@@ -720,47 +935,64 @@ def _fit(  # noqa: C901,PLR0912
...
 
-        logger.info("--- ✅ Fine-tuning Finished ---")
+        if is_main_process:
+            logger.info("--- ✅ Fine-tuning Finished ---")
 
         self._setup_inference_model(final_inference_eval_config)
```
Comment (anuragg1209) [inline src/tabpfn/finetuning/finetuned_base.py]: Minor: `_setup_inference_model()` runs on all ranks. Only rank 0 uses the inference model for prediction. Should we consider guarding with `if is_main_process`: to avoid redundant model cloning + fitting on non-main ranks ?

File: examples/finetune_classifier.py [hunk 3]
```diff
@@ -1,9 +1,13 @@
 """Example of fine-tuning a TabPFN classifier using the FinetunedTabPFNClassifier wrapper.
 
 Note: We recommend running the fine-tuning script on a CUDA-enabled GPU with 80 GB of VRAM.
+
+Multi-GPU: torchrun --nproc-per-node=N examples/finetune_classifier.py
```
Comment (anuragg1209) [inline examples/finetune_classifier.py:5]: Minor: I noticed there are no tests for the DDP finetuning code path (existing `test_finetuning_classifier.py` and `test_finetuning_regressor.py` don't cover distributed). What do you think about adding at least a basic smoke test? I understand DDP tests are tricky to run in CI (need multi-GPU), but maybe we could add a test that at least validates the setup/cleanup logic with mocked dist calls, or mark them as `@pytest.mark.skipUnless(torch.cuda.device_count() > 1)`. Happy to add the tests if you think it makes sense. Open to your thoughts!

File: src/tabpfn/finetuning/finetuned_base.py (hunk 1 above)
Comment (psinger-prior) [reply src/tabpfn/finetuning/finetuned_base.py:992]: I would say this is an extreme edge case that is not really common, I still added a note to the examples.

File: examples/finetune_classifier.py (hunk 3 above)
Comment (psinger-prior) [reply examples/finetune_classifier.py:5]: I think mocked tests are not too useful here, we would need real tests on multi-gpu runners. I would rely on the manual testing for now and fix issues along the way as this is also not core functionality of the package.

Comment (anuragg1209) [reply examples/finetune_classifier.py:5]: Sounds good!
